### DISCUSSION TITLE
How can I facet on multiple variables?

### DISCUSSION THREAD

**Discussion by rongcuid** [vega/altair#3601] [Q&A] [answered]
I tried to figure out by reading examples/docs of both Altair and Vega-Lite, but I don't find any. What I am trying to achieve is the following equivalent in ggplot2:

```
facet_wrap(["V1", "V2", "V3"])
```

However, I can't find anyway to make Altair `.facet` accept multiple fields. There seem to be some mention about nested facets in Vega-Lite Github issues, but I can't find if it's actually implemented or how to use them.

Any suggestions?

**Comment by joelostblom** [comment]
You could use `row` and `column` like in this example https://altair-viz.github.io/gallery/bar_chart_faceted_compact.html. If you would like to include more variables than two, I think you would have to create new combined columns of the variable combinations first. But I'm not 100% on how ggplot handles it, so if this does not sound like what you mean, feel free to include a short example with complete ggplot code and output.

**Reply by rongcuid** [reply]
I am specifically trying to facet with more than 2 variables.

Example using Plotnine:

```py
# %%
from plotnine import *
from plotnine.data import mpg

(
    ggplot(mpg, aes("cty", "hwy"))
    + geom_point(size=.5)
    + facet_wrap(["manufacturer", "model", "drv"])
    + theme(figure_size=(6, 6))
)
```
![cars](https://github.com/user-attachments/assets/0aa1516a-83d6-4ce1-a948-6331553df47e)

**Reply by joelostblom** [reply, marked as answer]
There is currently no support for that in Vega-Lite so it is not available in Altair. Instead, you can create the new combined column with the dataframe library you're using, or directly in altair using `transform_calculate`:

```py
import pandas as pd


mpg = pd.read_csv('https://raw.githubusercontent.com/has2k1/plotnine/4159f9548fa5f95c02366efc324e88f9ba79af09/plotnine/data/mpg.csv')

alt.Chart(mpg, width=100, height=100).mark_point().transform_calculate(
    man_model_drv=alt.expr.slice([alt.datum.manufacturer, alt.datum.model, alt.datum.drv])
    # Equivalent JS string expression
    # man_model_drv='[datum.manufacturer, datum.model, datum.drv]'
).encode(
    x='cty',
    y='hwy',
).facet(
    'man_model_drv:N',
    columns=7
)
```
![image](https://github.com/user-attachments/assets/3bb80704-1c7b-4c83-ae34-f9a1a45531cd)

---

Initial suggestion before switching to creating the array directly:

```
    # the call to `split` is to create the newline in the title
    man_model_drv=alt.expr.split(alt.datum.manufacturer + "-" + alt.datum.model + "-" + alt.datum.drv, '-')
```

**Reply by dangotbanned** [reply]
@joelostblom do we have an example like this in the gallery? 

I'm somewhat blown away by how quickly you recreated https://github.com/vega/altair/discussions/3601#discussioncomment-10698021

**Reply by joelostblom** [reply]
Not yet, but I could maybe append on to https://altair-viz.github.io/user_guide/compound_charts.html#faceted-charts? I was planning to update that section with some other changes anyways.

**Reply by rongcuid** [reply]
Can I just join them with `\n` or something? Some of my variables have `-` in it, and I would like to avoid picking some special character every time.

**Reply by joelostblom** [reply]
You can join with any character that `split` accepts, since you need an array for text to render as multi line. I'm not sure on the top of my head if newline is supported, but try it it out.

**Reply by rongcuid** [reply]
Why do I need to make a string and then split? Is it impossible to create an array directly?

**Reply by joelostblom** [reply]
Good suggestion. I didn't think it was possible, but it turns out you can just do this inside the transform:

```py
man_model_drv=alt.expr.slice([alt.datum.manufacturer, alt.datum.model, alt.datum.drv])
```

or the following(see https://github.com/vega/altair/issues/3605 for details of the differences between these two approaches)

```py
man_model_drv='[datum.manufacturer, datum.model, datum.drv]'
```

I updated my answer above with this
